### DISCUSSION TITLE
Dakota OUU primary and secondary response mapping

### DISCUSSION THREAD

**Discussion by rameshthar9** [snl-dakota/dakota#49] [Q&A] [answered]
Hello, Dakota community,
I am new to Dakota. I modified the example nested OUU to work with Matlab in Windows. I ran a simple cantilever model and got an output file, but I couldn't entirely understand a few things in this file. I couldn't understand how the below mapping works. I appreciate it if someone could explain what each number is referring to.
primary_response_mapping   = 1. 0. 0. 0. 0. 0.
secondary_response_mapping = 0. 0. 1. 3. 0. 0.
                                                       0. 0. 0. 0. 1. 3.

Another question is, there is no reference as to which variables are associated with these three response functions. How does Dakota know which is which?

In the OPTImization response:
id_responses = 'OPTIM_R'
	objective_functions = 1
	nonlinear_inequality_constraints = 2
	analytic_gradients
	no_hessians

In the UQ response:
responses,
	id_responses = 'UQ_R'
	response_functions = 3
	no_gradients
	no_hessians

Are the UQ response functions referring to the same response functions as OPTIM_R, and how do you know which is the objective function and which is the nonlinear_inequality_constraints are for?
Please see below is the".IN" file I used.

[dakota_ouu1_cantilevertxt.txt](https://github.com/snl-dakota/dakota/files/12149433/dakota_ouu1_cantilevertxt.txt)

Thank you in advance.

#@ s*: Label=FastTest
#@ *: DakotaConfig=HAVE_DOT
# DAKOTA Input File: dakota_ouu1_cantilever.in
# Optimization under uncertainty
# using the cantilever test function.

environment,
	  method_pointer = 'OPTIM'
#TR added table_data 
	  tabular_data
    tabular_data_file = 'cantilever_centered_tabular.dat'

###########################
# begin opt specification #
###########################
method,
	id_method = 'OPTIM'
	model_pointer = 'OPTIM_M'
	#dot_sqp
	conmin_frcg
#	npsol_sqp
	  convergence_tolerance = 1.e-6
	scaling

model,
	id_model = 'OPTIM_M'
	nested
	  variables_pointer  = 'OPTIM_V'
	  sub_method_pointer = 'UQ'
	  responses_pointer  = 'OPTIM_R'
# Approach 1: constrain the probabilities < .00135
#	  primary_response_mapping   = 1. 0. 0. 0. 0. 0. 0. 0. 0.
#	  secondary_response_mapping = 0. 0. 0. 0. 0. 1. 0. 0. 0.
#                                      0. 0. 0. 0. 0. 0. 0. 0. 1.
# Approach 2 (better for sampling): constrain mean + 3 sigma
	  primary_response_mapping   = 1. 0. 0. 0. 0. 0.
	  secondary_response_mapping = 0. 0. 1. 3. 0. 0.
                                       0. 0. 0. 0. 1. 3.

variables,
	id_variables = 'OPTIM_V'
	continuous_design = 2
	  initial_point    2.5    2.5
	  upper_bounds    10.0   10.0
	  lower_bounds     1.0    1.0
	  descriptors 'w' 't'
	scale_type = 'auto'

responses,
# minimize mean Weight
# s.t.     p_fail_S <= .00135 (beta = 3.0) [Approach 1]
#          p_fail_D <= .00135 (beta = 3.0) [Approach 1]
# or
#          mean_S + 3 sigma_S <= 0         [Approach 2]
#          mean_D + 3 sigma_D <= 0         [Approach 2]
#
# NOTE: This specifies the TOTAL RESPONSE for the optimization,
#       which is a combination of nested & interface responses.
	id_responses = 'OPTIM_R'
	objective_functions = 1
	nonlinear_inequality_constraints = 2
# Approach 1 allowables
#	nonlinear_inequality_upper_bounds = .00135 .00135
# Approach 2 uses default upper bounds (mean + 3 sigma < 0.0)
	numerical_gradients
	  method_source dakota
	  interval_type central
	  fd_gradient_step_size = 1.e-2
	no_hessians

##########################
# begin UQ specification #
##########################
method,
	id_method = 'UQ'
	model_pointer = 'UQ_M'
        sampling
	  samples = 10 sample_type lhs
	  seed = 12347 
#	  fixed_seed					#s1
	  output quiet

model,
	id_model = 'UQ_M'
	single
	  variables_pointer = 'UQ_V'
	  interface_pointer = 'UQ_I'
	  responses_pointer = 'UQ_R'

variables,
	id_variables = 'UQ_V'
	continuous_design = 2
	normal_uncertain = 4
	  means             =  40000. 29.E+6 500. 1000.
	  std_deviations    =  2000. 1.45E+6 100. 100.
	  descriptors       =  'P' 'E' 'X' 'Y'

interface,
	id_interface = 'UQ_I'
	fork,
	parameters_file = 'params.in'		
	results_file    = 'results.out'		

		analysis_driver = 'cscript matlab_cantilever_bb_simulator.vbs'	
	  deactivate evaluation_cache restart_file

responses,
	id_responses = 'UQ_R'
	response_functions = 3
	no_gradients
	no_hessians

**Comment by jadamstephens** [comment]
Please take a look at the keyword documentation for [sub_method_pointer](https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/reference/model-nested-sub_method_pointer.html) and its subkeywords. They explain how Dakota uses the primary and secondary response mapping matrices to compute responses for the outer method and model from results computed by the inner method.

> Another question is, there is no reference as to which variables are associated with these three response functions. How does Dakota know which is which?

Dakota does not associate variables with specific objective functions or nonlinear constraints. The variables should be the superset needed to compute all the objective functions. It's up to your driver to compute each objective function and nonlinear constraint using the required subset of the variables.

 > Are the UQ response functions referring to the same response functions as OPTIM_R, and how do you know which is the objective function and which is the nonlinear_inequality_constraints are for?

No. Hopefully reading the reference documentation will clarify this, but I'll also say a little about it here. In a nested study such as OUU, the responses associated with the UQ method (those defined in `UQ_R`) come from your analysis driver. The responses associated with the optimization method (objective functions and, optionally, nonlinear constraints), in `OPTIM_R`, are computed from the results of the nested UQ study by using the primary and secondary response mappings.

Think of a nested study like one loop nested within another. The optimization method is the outer loop, and the UQ method is the inner loop. For each iteration of the optimization:
1.  Dakota inserts values of optimization variables into the UQ study. These values are frozen for the duration of the UQ study. The other UQ variables are varied.
2. A complete UQ study is run to produce results like the mean and standard deviation of each response.
3. These results are mapped into the objective functions and nonlinear constraints that belong to the optimizer using the mapping matrices.

**Reply by rameshthar9** [reply]
I really appreciate your help. I have started to understand how the mapping works, but little confused as to the mapping position.

This is the example from the sub-method pointer section in the reference manual.
 
primary_response_mapping  0 0 0.3 0 0 0.7

This translates into:
Sum_ p = 0.3 * ( 90th percentiles of f1) + 0.7 * ( 90th percentiles of f2) 

Here f1 is the third one in the mapping, and f2 is in the 6th position, but only two responses come from the inner loop. So, what are the first two?

**Reply by jadamstephens** [reply, marked as answer]
The mean and standard deviation of f1. The 4th and 5th values in the mapping are the multipliers for the mean and standard deviation of f2.
